src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+          new SyntaxError(`The element name '${name}' is not valid.`));
+      }
+      if (this._definitions.has(name)) {
+        return Promise.resolve();
+      }
+      var deferred = {
+        promise: null,
+      };
+      deferred.promise = new Promise(function(resolve, _) {
+       deferred.resolve = resolve;
+      });
+      this._whenDefinedMap.set(name, deferred);
+      return deferred.promise;
+    },
+
+    /**
+     * Causes all pending mutation records to be processed, and thus all
+     * customization, upgrades and custom element reactions to be called.
+     * `enableFlush` must be true for this to work. Only use during tests!
+     */
+    flush: function() {
+      if (this.enableFlush) {
+        console.warn("flush!!!");
+        this._observers.forEach(function(observer) {
+          this._handleMutations(observer.takeRecords());
+        }, this);
+      }
+    },
+
+    _setNewInstance: function(instance) {
+      this._newInstance = instance;
+    },
+
+    /**
+     * Observes a DOM root for mutations that trigger upgrades and reactions.
+     * @private
+     */
+    _observeRoot: function(root) {
+      root.__observer = new MutationObserver(this._handleMutations.bind(this));
+      root.__observer.observe(root, {childList: true, subtree: true});
+      if (this.enableFlush) {
+        // this is memory leak, only use in tests
+        this._observers.add(root.__observer);
+      }
+    },
+
+    /**
+     * @private
+     */
+    _unobserveRoot: function(root) {
+      if (root.__observer) {
+        root.__observer.disconnect();
+        root.__observer = null;
+        if (this.enableFlush) {
+          this._observers.delete(root.__observer);
+        }
+      }
+    },
+
+    /**
+     * @private
+     */
+    _handleMutations: function(mutations) {
+      for (var i = 0; i < mutations.length; i++) {
+        var mutation = mutations[i];
+        if (mutation.type === 'childList') {
+          // Note: we can't get an ordering between additions and removals, and
+          // so might diverge from spec reaction ordering
+          this._addNodes(mutation.addedNodes);
+          this._removeNodes(mutation.removedNodes);
+        }
+      }
+    },
+
+    /**
+     * @param {NodeList} nodeList
+     * @private
+     */
+    _addNodes: function(nodeList) {
+      for (var i = 0; i < nodeList.length; i++) {
+        var root = nodeList[i];
+
+        if (!isElement(root)) {
+          continue;
+        }
+
+        // Since we're adding this node to an observed tree, we can unobserve
+        this._unobserveRoot(root);
+
+        var walker = createTreeWalker(root);
+        do {
+          var node = /** @type {HTMLElement} */ (walker.currentNode);
+          var definition = this._definitions.get(node.localName);
+          if (definition) {
+            if (!node.__upgraded) {
+              this._upgradeElement(node, definition, true);
+            }
+            if (node.__upgraded && !node.__attached) {
```

Comment:
Is the node guaranteed to be in the document at this point? If not, then should not call `connectedCallback`. For example, this should (probably) not be called in HTMLImports.